pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java

```diff
@@ -183,4 +194,26 @@ public GroupKeyGenerator getGroupKeyGenerator() {
   public GroupByResultHolder[] getGroupByResultHolders() {
     return _groupByResultHolders;
   }
+
+  private Integer getGroupByResultHolderCapacityBasedOnFilterPredicate(QueryContext queryContext) {
+    if (queryContext.getFilter() == null || queryContext.getGroupByExpressions() == null) {
+      return null;
+    }
+
+    List<FilterContext> filterContexts = queryContext.getFilter().getChildren() != null
+        ? queryContext.getFilter().getChildren()
+        : List.of(queryContext.getFilter());
+
+    Map<ExpressionContext, InPredicate> predicateMap = filterContexts.stream()
+        .map(FilterContext::getPredicate)
+        .filter(predicate -> predicate.getType() == Predicate.Type.IN)
+        .collect(Collectors.toMap(Predicate::getLhs, predicate -> (InPredicate) predicate));
+
+    OptionalInt result = queryContext.getGroupByExpressions().stream()
```

Comment:
If there is a single groupBy expression with a filter on the groupBy, this is correct. 

However, consider the case where there are 2 groupByExpressions but only a filter on 1 column. In this case this optimization will not work.

Reply:
handled this case